### PR TITLE
GEODE-6887: Add LegacyStatCounter and LegacyStatTimer

### PR DESCRIPTION
**Pull request by demery-pivotal**
Added `LegacyStatCounter` and `LegacyStatTimer`. Updated `CacheServerStats` to use a `LegacyStatCounter` to report "events received."

For now, `LegacyStatCounter` allows binding to int stats, so that we can use it for our existing gateway receiver events received meter. After all int stats are converted to long stats (in a separate PR), we will remove the ability to bind to int stats.

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [X] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [X] Is your initial contribution a single, squashed commit?

- [X] Does `gradlew build` run cleanly?

- [X] Have you written or updated unit tests to verify your changes?

Please review @kirklund @pivotal-jbarrett @upthewaterspout @moleske @aaronlindsey @mhansonp

### CODE REVIEW

File: geode-core/src/test/java/org/apache/geode/internal/cache/wan/GatewayReceiverStatsTest.java [hunk 1]
```diff
@@ -60,27 +63,50 @@ public void setup() {
...
     int delta = 99;
 
     gatewayReceiverStats.incEventsReceived(delta);
 
     verify(statistics).incInt(eventsReceivedStatId, delta);
+  }
 
-    gatewayReceiverStats.close();
+  @Test
+  public void incEventsReceived_incrementsTheRegisteredEventsReceivedCounter() {
```
Comment (aaronlindsey) [inline geode-core/src/test/java/org/apache/geode/internal/cache/wan/GatewayReceiverStatsTest.java]: The first 5 tests in this class seem to be testing the same things that are already tested in LegacyStatCounterTest, i.e. making sure increment increments both the stat and the counter, that the counter returns the value from the stat, and that the description and units on the counter match the stat. Rather than testing these same things for each LegacyStatCounter we add, I would rather have these tests live in LegacyStatCounterTest.

File: geode-core/src/main/java/org/apache/geode/internal/statistics/meters/LegacyStatCounter.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+  @Override
+  public Id getId() {
+    return underlyingCounter.getId();
+  }
+
+  /**
+   * Returns a builder that can associate a stat with the eventual {@code LegacyStatCounter}. To
+   * associate the counter with a stat, call one of
+   * {@link Builder#doubleStatistic(Statistics, int)} or
+   * {@link Builder#longStatistic(Statistics, int)}.
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply geode-core/src/main/java/org/apache/geode/internal/statistics/meters/LegacyStatCounter.java]: Don't worry, the PR to remove int stats will beat this PR.

File: geode-core/src/main/java/org/apache/geode/internal/statistics/meters/IntStatisticBinding.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  }
+
+  @Override
+  public double doubleValue() {
+    return (double) statistics.getInt(statId);
+  }
+
+  @Override
+  public long longValue() {
+    return 0;
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply geode-core/src/main/java/org/apache/geode/internal/statistics/meters/IntStatisticBinding.java]: Good find! Though this will be superseded by #3713.

File: geode-core/src/main/java/org/apache/geode/internal/statistics/meters/LegacyStatCounter.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+      StatisticBinding statisticBinding) {
+    this.underlyingCounter = underlyingCounter;
+    this.statisticBinding = statisticBinding;
+  }
+
+  /**
+   * Increments both the underlying counter and the associated stat by the given amount.
+   */
+  @Override
+  public void increment(double amount) {
```
[the remark this answers is not in the document]
Comment (kirklund) [reply geode-core/src/main/java/org/apache/geode/internal/statistics/meters/LegacyStatCounter.java]: Geode statistics use both positive and negative amounts in increment calls. Not sure how this translates for meters, but maybe negative is ok.

File: geode-core/src/main/java/org/apache/geode/internal/statistics/meters/IntStatisticBinding.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (demery-pivotal) [reply geode-core/src/main/java/org/apache/geode/internal/statistics/meters/IntStatisticBinding.java]: Now that https://github.com/apache/geode/pull/3713 is in, I'll want to get rid of the int stuff before merging.

File: geode-core/src/main/java/org/apache/geode/internal/statistics/meters/LegacyStatCounter.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (demery-pivotal) [reply geode-core/src/main/java/org/apache/geode/internal/statistics/meters/LegacyStatCounter.java]: Though Micrometer describes counters as monotonically increasing, nothing in Micrometer enforces that. Given the lack of enforcement in Micrometer, I think I don't want to add that enforcement here.

File: geode-core/src/test/java/org/apache/geode/internal/cache/wan/GatewayReceiverStatsTest.java (hunk 1 above)
Comment (demery-pivotal) [reply geode-core/src/test/java/org/apache/geode/internal/cache/wan/GatewayReceiverStatsTest.java]: I do want to ensure that `GatewayReceiverStats` connects its `incEventsReceived()` and `getEventsReceived()` methods to the appropriate stat, which is something `LegacyStatCounter` can test on its own. The tests are similar, but there's a key difference in the responsibilities being tested: The `LegacyStatCounter` tests test that the counter connects to whatever stat you tell it to connect to. The `GatewayReceiverStats` tests test that these particular `GatewayReceiverStats` methods connect specifically to the "events received" stat.

File: geode-core/src/main/java/org/apache/geode/internal/statistics/meters/IntStatisticBinding.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (demery-pivotal) [reply geode-core/src/main/java/org/apache/geode/internal/statistics/meters/IntStatisticBinding.java]: On second thought, I think we'll remove that stuff in a separate PR.

**Comment by demery-pivotal** [comment]
I rebased to pick up https://github.com/apache/geode/pull/3713, to make sure this PR works with those changes.

File: geode-core/src/main/java/org/apache/geode/internal/statistics/meters/LegacyStatCounter.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply geode-core/src/main/java/org/apache/geode/internal/statistics/meters/LegacyStatCounter.java]: Like Micrometer, Geode distinguished at the API level between counters and gauges but both don't enforce positive increments only for counters. A Micrometer registry may enforce this though so we should not use `Counters` when we want a `Gauge`.

[the remark this answers is not in the document]
Comment (demery-pivotal) [reply geode-core/src/main/java/org/apache/geode/internal/statistics/meters/LegacyStatCounter.java]: Yes. I'm not sure yet what kind of meter we'll use for gauge stats.

[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply geode-core/src/main/java/org/apache/geode/internal/statistics/meters/LegacyStatCounter.java]: Well, I would hope the `Gauge` meter. We don't have many so in those cases it may make sense to do what is best. Likely in those cases the service being gauged should probably have method for getting the current observed value and both the Micrometer `Meter` and legacy `Statistics` `LongSupplier` should be setup to poll that method.

[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply geode-core/src/main/java/org/apache/geode/internal/statistics/meters/LegacyStatCounter.java]: Many of these might be getting reworked like this already as a result of some of the performance work we are currently working on. Often these gauges were being used to determine behavior of the service it gauged. In the current implementation the service inc/dec some Statistics gauge and then read it when performing some action. This leads to poor performance since the read of `Statistics` is costly. It also leads to a bug when `Statistics` is turned off and the no-op implementation is in use, now all gauges are 0. With Micrometer in the mix the same issues would exist. We are moving all these uses back into the service. The service will keep its own gauge and the `Statistics` will poll it via a `LongSupplier`. The service will not need to poll the value from outside its own objects.

Long story short, we should consider all this and refactor some things when we hit gauges like this doing Micrometer stuff.
